### PR TITLE
Regions with dict

### PR DESCRIPTION
**Pull request by asbjrnmunk**

### CODE REVIEW

File: yucca/pipeline/managers/YuccaManager.py [hunk 1]
```diff
@@ -163,7 +163,7 @@ def initialize(
 
         seed_config = seed_everything_and_get_seed_config(ckpt_seed=self.ckpt_config.ckpt_seed)
 
-        plan_config = self.get_plan_config(
```
Comment (Sllambias) [inline yucca/pipeline/managers/YuccaManager.py:166]: This should be reverted, as this allows other managers to intervene how they get the plan config. if e.g. they want to manually specify the PlanConfig, without having to overwrite the whole initialize method

**Comment by asbjrnmunk** [comment]
Ready for review @Sllambias 🚀

File: yucca/pipeline/managers/YuccaManager.py [hunk 2]
```diff
@@ -252,6 +257,7 @@ def initialize(
         self.verify_modules_are_valid()
 
         self.trainer = L.Trainer(
+            accelerator="cpu",
```
Comment (Sllambias) [inline yucca/pipeline/managers/YuccaManager.py:260]: should prob not be here ;)

File: yucca/pipeline/task_conversion/Task012_BraTS21.py [hunk 3]
```diff
@@ -89,9 +89,12 @@ def convert(path: str, subdir: str = "brats21/training_data"):
         target_imagesTr,
         target_imagesTs,
         ("FLAIR", "T1", "T1CE", "T2"),
-        labels={0: "background", 1: "necrotic tumor core", 2: "peritumoral edematous/invaded tissue", 3: "GD-enhancing tumor"},
-        regions_in_order=[[1, 2, 3], [1, 3], [3]],
-        regions_labeled=[1, 2, 3],
+        labels={0: "BG", 1: "NCR", 2: "ED", 3: "ET"},
```
Comment (Sllambias) [inline yucca/pipeline/task_conversion/Task012_BraTS21.py:92]: I get how this is nice for the regions, but I'm not a huge fan of the string matching here for two reasons:
1. It's important to have "1: "necrotic tumor core", 2: "peritumoral edematous/invaded tissue", 3: "GD-enhancing tumor"" so we can actually see what our labels mean, and
2. It might be quite annoying and fragile to match super long strings, which I think there should be space for

File: yucca/pipeline/task_conversion/Task012_BraTS21.py [hunk 4]
```diff
@@ -89,9 +89,12 @@ def convert(path: str, subdir: str = "brats21/training_data"):
...
+        regions={
+            "WT": {"priority": 3, "labels": ["ET", "NCR", "ED"]},
```
Comment (Sllambias) [inline yucca/pipeline/task_conversion/Task012_BraTS21.py:94]: maybe it could become             "WT": {"priority": 3, "labels": [1, 2, 3]},
so they match the keys instead of the values

File: yucca/modules/lightning_modules/YuccaLightningModule_onehot_labels.py [hunk 5]
```diff
@@ -74,6 +76,12 @@ def training_step(self, batch, batch_idx):
     def validation_step(self, batch, batch_idx):
         inputs, target, file_path = batch["image"], batch["label"], batch["file_path"]
         output = self(inputs)
+
+        print("inputs", inputs.shape)
+        print("outputs", output.shape)
+        print("target", target.shape)
+        print("num classes", self.num_classes)
+
```
Comment (Sllambias) [inline yucca/modules/lightning_modules/YuccaLightningModule_onehot_labels.py:84]: prob shouldn't be here

File: yucca/functional/transforms/label_transforms.py [hunk 6]
```diff
@@ -1,32 +1,48 @@
 import numpy as np
-import torch
 
 
-def convert_labels_to_regions(data, regions):
-    b, c, *shape = data.shape
-    assert c == 1, "# Channels is not 1. Make sure the input to this function is a segmentation map of dims (b,c,h,w[,d])"
+def translate_region_labels(regions: dict, labels: dict) -> dict:
```
Comment (Sllambias) [inline yucca/functional/transforms/label_transforms.py:4]: this would depend on how we sort my other comments

File: yucca/functional/utils/torch_utils.py [hunk 7]
```diff
@@ -14,7 +14,7 @@ def get_available_device():
     if torch.cuda.is_available():
         return "cuda"
     elif torch.backends.mps.is_available():
-        return "mps"
+        return "cpu"  # we cannot use mps, since Conv3D is not supported here.
```
Comment (Sllambias) [inline yucca/functional/utils/torch_utils.py:17]: we can, just not for 3D, so this should be reverted

File: yucca/modules/lightning_modules/YuccaLightningModule_onehot_labels.py (hunk 5 above)
Comment (Sllambias) [inline yucca/modules/lightning_modules/YuccaLightningModule_onehot_labels.py:84]: :)

**Review by Sllambias — CHANGES_REQUESTED** [review]
have some requested changes

File: yucca/functional/utils/torch_utils.py (hunk 7 above)
Comment (asbjrnmunk) [reply yucca/functional/utils/torch_utils.py:17]: But tests are in 3D, and this means tests doesn't run locally.

File: yucca/pipeline/task_conversion/Task012_BraTS21.py (hunk 4 above)
Comment (asbjrnmunk) [reply yucca/pipeline/task_conversion/Task012_BraTS21.py:94]: My reasoning is this:

1. I wanted something where it is easy to translate a challenge submission to a region specification.
2. Label values are dependent on the task conversion file in e.g. BraTS.

I would say either are equally fragile. However, the asserts here, make sure that the long string matching will never be a problem, since this will be caught when task is converted: 

https://github.com/Sllambias/yucca/blob/4b11081affe14aa806f8fe6339f0aed044b96a1e/yucca/pipeline/task_conversion/utils.py#L117-L132

File: yucca/functional/utils/torch_utils.py (hunk 7 above)
Comment (Sllambias) [reply yucca/functional/utils/torch_utils.py:17]: Hmm, fair point and I guess this doesn't mess with manually setting training to MPS since we removed the device calls in training - and who would actually like to do things on MPS anyway tbh

File: yucca/pipeline/managers/YuccaManager.py (hunk 2 above)
Comment (asbjrnmunk) [reply yucca/pipeline/managers/YuccaManager.py:260]: Will change.

File: yucca/pipeline/managers/YuccaManager.py (hunk 1 above)
Comment (asbjrnmunk) [reply yucca/pipeline/managers/YuccaManager.py:166]: I dont think it makes sense to challenge this, but just want to state that i find that extremely bad practice.

**Comment by asbjrnmunk** [comment]
@Sllambias

File: yucca/pipeline/managers/YuccaManager.py [hunk 8]
```diff
@@ -163,6 +163,7 @@ def initialize(
 
         seed_config = seed_everything_and_get_seed_config(ckpt_seed=self.ckpt_config.ckpt_seed)
 
+        # WARNING: This might be overwritten by a subclass!
```
Comment (Sllambias) [inline yucca/pipeline/managers/YuccaManager.py:166]: I'd prefer to not have these kind of comments here, they seem a bit snarky ;)

**Review by Sllambias — APPROVED** [review]
I think you're right about the string matching - and it is kinda nice to see which semantic areas make a region, rather than just digits that you need to crossreference so let's go with your solution.
I'd just like the one comment to be removed and then LGTM :))
